source/h2_connection.c

```diff
@@ -1111,6 +1120,51 @@ static struct aws_h2err s_decoder_on_window_update(uint32_t stream_id, uint32_t
     return AWS_H2ERR_SUCCESS;
 }
 
+struct aws_h2err s_decoder_on_goaway_begin(
+    uint32_t last_stream,
+    uint32_t error_code,
+    uint32_t debug_data_length,
+    void *userdata) {
+    (void)debug_data_length;
+    struct aws_h2_connection *connection = userdata;
+
+    if (last_stream > connection->thread_data.goaway_received_last_stream_id) {
+        CONNECTION_LOGF(
+            ERROR,
+            connection,
+            "Peer GOAWAY frame has invalid last stream id %" PRIu32 ", which is higher than the one we received",
+            last_stream);
+        return aws_h2err_from_h2_code(AWS_H2_ERR_PROTOCOL_ERROR);
+    }
+    /* stop sending any new stream and making new request */
+    aws_atomic_store_int(&connection->synced_data.new_stream_error_code, AWS_ERROR_HTTP_GOAWAY_RECEIVED);
+    connection->thread_data.goaway_received = true;
+    connection->thread_data.goaway_received_error_code = error_code;
```

Comment:
I am not reporting the error at s_handler_shutdown(), since we will report it when we receive the GOAWAY. So I just feel maybe we can just report it here, and it seems more clear.

Reply:
So, I feel like the connection->thread_data.goaway_received_error_code seems serving no purpose now. Should I delete it?